src/eval/mod.rs

```diff
@@ -498,11 +498,20 @@ where
                     }
                 }
             }
-            Term::RecRecord(ts, dyn_fields, attrs) => {
+            Term::RecRecord(ts, dyn_fields, attrs, free_vars) => {
                 let rec_env = fixpoint::rec_env(ts.iter(), &env)?;
 
-                ts.iter()
-                    .try_for_each(|(_, rt)| fixpoint::patch_field(rt, &rec_env, &env))?;
+                let filter_env = |(id, _): &(Ident, Thunk)| -> bool {
+                    if let Some(free_vars) = free_vars.as_ref().and_then(|fr| fr.get(id)) {
+                        free_vars.contains(id)
```

Comment:
I don't think this is what we want. Here the logic of should we add field `id` to the environment of field `foo` doesn't depend on the field `foo` being patched, which is fishy.

You should have two `id`s: the one of the field we patch, and the one we are currently filtering on. That is, the filter function should be different for each field:

```
let filter_env = |field: Ident| {
  if let Some(free_vars) = free_vars.as_ref() {
    |(id, _)| free_vars.contains(id);
  }
  else {
    |_| true
  }
};
...
ts.iter().try_for_each(|(id, rt)| {
    fixpoint::patch_field(rt, &rec_env, &env, filter_env(id))
})?;
```

(May need type annotations or some `fn` instead of closures with additional parameters, but you get the idea)

Reply:
This may or may not explain why my counter-example below `{x = 1, let y = let x = x + 1 in x}` still passes.